### PR TITLE
[FEATURE][CORE] Introduce Instant Session Start

### PR DESCRIPTION
**Pull request by stefaus**
This is the same Patch as #128, which was reverted because #127 got reverted.
---
This patch implements the core functionality of Instant Session Start,
which is optional selectable for Project Negotiation. The main scope of
this feature is to decrease the waiting time for users, till the first
files are provided and they can participate in a session.

Deviant to previous approaches, this one uses a file stream based
solution. Compared to using Activities, this provides increased
performance, because of small overheads and better compresion
possibilities, big files support and is for now, not interfering with
existing activity handling code.

This first Version implements a basic prioritisation, by first sending
important eclipse project files, afterwards allways opened files first
and remaining sorted by path hierarchy.

Following features are excluded, to limit this rather big patch or are
planned for the future:

 - Proper handling of Activities, to remove editor lock

 - Compression (using a libary like zstd)

 - Configurable File Prioritisations

 - More progress infos (bandwith, remaining file size, time...)

 - Performance optimizations, refactoring of ProjectNegotiation

 - User interface enhancements

 - Priority file sending, while active file transmission

### CODE REVIEW

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        @Override
+        public void editorActivated(User user, SPath file) {
+            fileOpened(file);
+        }
+    };
+
+    private List<StartHandle> stoppedUsers = null;
+    private User remoteUser = null;
+
+    public InstantOutgoingProjectNegotiation(final JID peer, //
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java]: Code Smell Code Smell: Constructor has 10 parameters, which is greater than 7 authorized. ([squid:S00107](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS00107))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBe8anAXINBuRK2&open=AWUGszBe8anAXINBuRK2)

[ISSUE_KEY]: AWUGszBe8anAXINBuRK2

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+ * Receive shared Projects and display them instant using a stream based
+ * solution.
+ */
+public class InstantIncomingProjectNegotiation extends
+    AbstractIncomingProjectNegotiation {
+
+    private static final Logger log = Logger
+        .getLogger(InstantIncomingProjectNegotiation.class);
+
+    public InstantIncomingProjectNegotiation(
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java]: Code Smell Code Smell: Constructor has 11 parameters, which is greater than 7 authorized. ([squid:S00107](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS00107))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBl8anAXINBuRK4&open=AWUGszBl8anAXINBuRK4)

[ISSUE_KEY]: AWUGszBl8anAXINBuRK4

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        /*
+         * the user who sends this ProjectNegotiation is now responsible for the
+         * resources of the contained projects
+         */
+        for (Entry<String, IProject> entry : projectMapping.entrySet()) {
+            final String projectID = entry.getKey();
+            final IProject project = entry.getValue();
+
+            session.addProjectMapping(projectID, project);
+            /* TODO change queuing to resource based queuing */
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBl8anAXINBuRK3&open=AWUGszBl8anAXINBuRK3)

[ISSUE_KEY]: AWUGszBl8anAXINBuRK3

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    private void fileOpened(SPath file) {
+        if (file != null) {
+            openedFiles.addFirst(file);
+            log.debug(this + ": added " + file + " to open files queue");
+        }
+    }
+
+    private void sendProjectConfigFiles(OutgoingStreamProtocol osp)
+        throws IOException, LocalCancellationException {
+        /* TODO should be configurable in future */
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBe8anAXINBuRK1&open=AWUGszBe8anAXINBuRK1)

[ISSUE_KEY]: AWUGszBe8anAXINBuRK1

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        transmitter.send(ISarosSession.SESSION_CONNECTION_ID, getPeer(), //
+            StartActivityQueuingResponse.PROVIDER //
+                .create( //
+                new StartActivityQueuingResponse(getSessionID(), getID())));
+
+        checkCancellation(CancelOption.NOTIFY_PEER);
+
+        /* only get files, if something is missing */
+        int filesMissing = 0;
```
Comment (tobous) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java]: You could fix the naming thing (using `filesMissing` and `missingFiles` for completely different purposes) in this PR or in #156. It does not really matter to me.

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java (hunk 4 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBe8anAXINBuRK1&open=AWUGszBe8anAXINBuRK1)

[ISSUE_KEY]: AWUGszBe8anAXINBuRK1

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java (hunk 1 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java]: Code Smell Code Smell: Constructor has 10 parameters, which is greater than 7 authorized. ([squid:S00107](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS00107))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBe8anAXINBuRK2&open=AWUGszBe8anAXINBuRK2)

[ISSUE_KEY]: AWUGszBe8anAXINBuRK2

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java (hunk 3 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBl8anAXINBuRK3&open=AWUGszBl8anAXINBuRK3)

[ISSUE_KEY]: AWUGszBl8anAXINBuRK3

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java (hunk 2 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java]: Code Smell Code Smell: Constructor has 11 parameters, which is greater than 7 authorized. ([squid:S00107](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS00107))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=183&issues=AWUGszBl8anAXINBuRK4&open=AWUGszBl8anAXINBuRK4)

[ISSUE_KEY]: AWUGszBl8anAXINBuRK4

**Pull request merged**